File: Javascript/Algorithms/641.js
/**
 * Initialize your data structure here. Set the size of the deque to be k.
 * @param {number} k
 */
var MyCircularDeque = function(k) {
  this.values = Array(k);
  this.num_elements = 0;
  this.front = 0;
};

/**
 * Adds an item at the front of Deque. Return true if the operation is
 * successful.
 * @param {number} value
 * @return {boolean}
 */
MyCircularDeque.prototype.insertFront = function(value) {
  if (this.isFull()) return false;

  let st = this.values
           // new position of front will be front - 1 mod the size of deque.
           this.front = (this.front + st.length - 1) % st.length;
  st[this.front] = value;
  this.num_elements++;
  return true;
};

/**
 * Adds an item at the rear of Deque. Return true if the operation is
 * successful.
 * @param {number} value
 * @return {boolean}
 */
MyCircularDeque.prototype.insertLast = function(value) {
  if (this.isFull()) return false;

  // End (tail) is going to be front + length mod size of deque
  let back = (this.front + this.num_elements) % this.values.length;
  this.values[back] = value;
  this.num_elements++;
  return true;
};

/**
 * Deletes an item from the front of Deque. Return true if the operation is
 * successful.
 * @return {boolean}
 */
MyCircularDeque.prototype.deleteFront = function() {
  if (this.isEmpty()) return false;

  // to delete, just advance front by one mod size.
  this.front = (this.front + 1) % this.values.length;
  this.num_elements--;
  return true;
};

/**
 * Deletes an item from the rear of Deque. Return true if the operation is
 * successful.
 * @return {boolean}
 */
MyCircularDeque.prototype.deleteLast = function() {
  if (this.isEmpty()) return false;

  // affects get/insert{rear} which depend on it.
  this.num_elements--;
  return true;
};

/**
 * Get the front item from the deque.
 * @return {number}
 */
MyCircularDeque.prototype.getFront = function() {
  if (this.isEmpty()) return -1;
  return this.values[this.front]
};

/**
 * Get the last item from the deque.
 * @return {number}
 */
MyCircularDeque.prototype.getRear = function() {
  if (this.isEmpty()) return -1;

  // Calculate end based on front and length mod size. Note that
  // front + length denotes empty position in which we could insert
  // (see insertBack), this is why we decrement it by one.
  size = this.values.length;
  return this.values[(this.front + this.num_elements - 1) % size];
};

/**
 * Checks whether the circular deque is empty or not.
 * @return {boolean}
 */
MyCircularDeque.prototype.isEmpty = function() {
  return this.num_elements == 0;
};

/**
 * Checks whether the circular deque is full or not.
 * @return {boolean}
 */
MyCircularDeque.prototype.isFull = function() {
  return this.num_elements == this.values.length;
};

File: Python/Algorithms/641.py
class MyCircularDeque:

    class _node:
        def __init__(self, value, _next=None, _prev=None):
            self.value = value
            self.next = _next
            self.prev = _prev

    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        """
        self.size = k
        self.length = 0
        self.head = self.tail = None

    def insertFront(self, value: int) -> bool:
        """
        Adds an item at the front of Deque. Return true if the operation is successful.
        """
        if self.length >= self.size:
            return False

        if self.head:
            old = self.head
            self.head = self._node(value, old)
            old.prev = self.head
        else:
            self.head = self.tail = self._node(value)
        self.length += 1
        return True

    def insertLast(self, value: int) -> bool:
        """
        Adds an item at the rear of Deque. Return true if the operation is successful.
        """
        if self.length >= self.size:
            return False

        if self.tail:
            tmp = self.tail
            self.tail = self._node(value, None, tmp)
            tmp.next = self.tail
        else:
            self.head = self.tail = self._node(value)
        self.length += 1
        return True

    def deleteFront(self) -> bool:
        """
        Deletes an item from the front of Deque. Return true if the operation is successful.
        """
        if not self.length:
            return False

        self.length -= 1
        if self.length == 0:
            self.head = self.tail = None
            return True
        self.head = self.head.next
        self.head.prev = None
        return True

    def deleteLast(self) -> bool:
        """
        Deletes an item from the rear of Deque. Return true if the operation is successful.
        """
        if not self.length:
            return False

        self.length -= 1
        if self.length == 0:
            self.head = self.tail = None
            return True
        self.tail = self.tail.prev
        self.tail.next = None
        return True

    def getFront(self) -> int:
        """
        Get the front item from the deque.
        """
        if self.length == 0:
            return -1
        return self.head.value

    def getRear(self) -> int:
        """
        Get the last item from the deque.
        """
        if self.length == 0:
            return -1
        return self.tail.value

    def isEmpty(self) -> bool:
        """
        Checks whether the circular deque is empty or not.
        """
        return self.length == 0

    def isFull(self) -> bool:
        """
        Checks whether the circular deque is full or not.
        """
        return self.length == self.size
